Mindtpy ecp

soccercider opened the pull request:
## Fixes # .

## Summary/Motivation:


## Changes proposed in this PR:
-
-

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

soccercider committed:
First commit for Saeed to edit ECP

soccercider committed:
Changes initial_binary indentation and name in tests

soccercider committed:
Adding changes to branch for ECP initial working implementation

soccercider committed:
Merging conflict solved

soccercider committed:
remove name from ecp

bernalde commented on pyomo/contrib/mindtpy/initialization.py:
```diff
@@ -61,7 +61,7 @@ def MindtPy_initialize_master(solve_data, config):
         # if config.strategy == 'ECP':
         #     add_ecp_cut(solve_data, config)
```
This line here suggests that we were adding the cuts at the initial point for our implementation. This makes sense if we are dealing with a convex problem since at the initialization we can evaluate the variables. Consider uncommenting this line and modifying the code (updating `add_ecp_cut(solve_data, config)` to your new function `add_ecp_cuts(.)`)

bernalde commented on pyomo/contrib/mindtpy/MindtPy.py:
```diff
@@ -320,7 +320,9 @@ def solve(self, model, **kwds):
-            process_objective(solve_data, config, use_mcpp=False)
+
+            # Process objective in case that it is nonlinear move it to constraints
+            process_objective(solve_data, config)
```
We didn't notice that the merge removed the option `use_mcpp=False`, from this function. Please fix it.

bernalde commented on pyomo/contrib/mindtpy/cut_generation.py:
```diff
@@ -93,6 +93,50 @@ def add_oa_cuts(target_model, dual_values, solve_data, config,
+                linearize_violated=True,
+                linearize_inactive=False):
+    """Linearizes nonlinear constraints.
+
```
Please complete the documentation here to say that it adds the cuts for the ECP method.

bernalde commented on pyomo/contrib/mindtpy/cut_generation.py:
```diff
@@ -93,6 +93,50 @@ def add_oa_cuts(target_model, dual_values, solve_data, config,
+
+        constr_vars = list(identify_variables(constr.body))
+        jacs = solve_data.jacobians
+
```
This was extremely clever. I checked and the code above uses the duals only to determine which direction of the cut should be added if we have an equality constraint. The more I think about it, the more it makes sense to use the fact of `constr.uslack() < 0` and `constr.lslack() < 0` to determine which side of the cut should be added. What do you think @ZedongPeng ?

bernalde commented on pyomo/contrib/mindtpy/iterate.py:
```diff
@@ -172,6 +179,24 @@ def algorithm_should_terminate(solve_data, config, check_cycling):
+                    return False
+            if nlc.has_ub():
+                if nlc.uslack() < -config.ECP_tolerance:
+                    return False
```
Before the return could you have the logger reporting which constraint is still violated?

bernalde commented on pyomo/contrib/mindtpy/iterate.py:
```diff
@@ -172,6 +179,24 @@ def algorithm_should_terminate(solve_data, config, check_cycling):
+                if nlc.uslack() < -config.ECP_tolerance:
+                    return False
+        config.logger.info(
+            'MindtPy exiting on nonlinear constraints satisfaction. '
```
`Mindtpy-ECP`

bernalde reviewed (changes requested):
This is the first review and for instance, I have requested some changes. Make sure to address them in your next "pushes". You can also respond to the comments here.
It's really impressive how fast you got everything working!

soccercider committed:
These are in response to David's comments.
initialization.py: added logging on initial strategy chosen
MindtPy.py: fixed process_objective(solve_data, config, use_mcpp=False)
iterate.py: added logging on which nonlinear constraint fails for ecp. Trying to fix overflow error but need help, along with quick check of ecp in integer cut section (2 lines commented)
cut_generation.py: attempting to fix overflow error, but not fixed. The following errors show up:
ERROR: evaluating object as numeric value: exp(X[22])
OverflowError: math range error

Upon inspection, the math error occurs where most of the continuous variables go to 0, whereas a select few approach inf. Taking exp of that large number causes overflow.

Addiditonally, in ECP there is no NLP  subproblem, so there is no UB to help with convergence.

bernalde commented on pyomo/contrib/mindtpy/cut_generation.py:
```diff
@@ -124,9 +146,9 @@ def add_ecp_cuts(target_model, solve_data, config,
-                or (linearize_inactive and constr.lslack() > 0):
+            and (linearize_active and abs(my_lslack) < config.ECP_tolerance) \
+                or (linearize_violated and my_lslack < 0) \
+                or (linearize_inactive and my_lslack > 0):
```
Maybe change these variable names. Try using PyCharms "rename" tools. Also, remove the intermediate print statements for the commits.

ZedongPeng commented on pyomo/contrib/mindtpy/cut_generation.py:
```diff
@@ -93,6 +93,72 @@ def add_oa_cuts(target_model, dual_values, solve_data, config,
+            )
+        #print("uslack is " , constr.body())
+        #print("lslack is " ,constr.upper())
+        if constr.has_ub() \
```
Do we need to consider the nonlinear equality constraint case here?

bernalde replied:
We are in the process of figuring that out. For equality constraints, we need to figure out which direction of the cut is going to be added and that was done using the duals from the NLP (unavailable for ECP).

ZedongPeng commented on pyomo/contrib/mindtpy/iterate.py:
```diff
@@ -28,7 +31,7 @@ def MindtPy_iteration_loop(solve_data, config):
         solve_data.mip_subiter = 0
         # solve MILP master problem
-        if config.strategy == 'OA':
+        if config.strategy == 'OA' or config.strategy == 'ECP':
```
`if config.strategy in {'OA', 'ECP'}:`

ZedongPeng commented on pyomo/contrib/mindtpy/initialization.py:
```diff
@@ -53,23 +55,29 @@ def MindtPy_initialize_master(solve_data, config):
-                                                    solve_data, config)
+        if config.strategy == 'ECP':
+            add_ecp_cuts(solve_data.mip, solve_data, config)
+    elif config.init_strategy == 'initial_binary':
```
Is this the bug you mentioned that Saeed found in initialization.py? @bernalde

ZedongPeng commented on pyomo/contrib/mindtpy/initialization.py:
```diff
@@ -53,23 +55,29 @@ def MindtPy_initialize_master(solve_data, config):
+            add_ecp_cuts(solve_data.mip, solve_data, config)
+    elif config.init_strategy == 'initial_binary':
+        if config.strategy != 'ECP':
+            fixed_nlp, fixed_nlp_result = solve_NLP_subproblem(solve_data, config)
```
I am a little curious about whether `TransformationFactory('core.fix_integer_vars').apply_to(fixed_nlp)` still works if a initial value has been provided. Or have we already got the value for each variable before this line?

bernalde replied:
It does work since the transformation uses the current variable value to fix the values. If no initial value has been provided, we rather "fail" and inform the user that a given variable is missing an initial value. Let's talk today on how to deal with that.

bernalde commented on pyomo/contrib/mindtpy/iterate.py:
```diff
@@ -184,14 +187,32 @@ def algorithm_should_terminate(solve_data, config, check_cycling):
+            try:
+                my_uslack = nlc.uslack()
+            except OverflowError:
+                my_uslack = 1e10
```
Same issue with renaming `my_uslack ` and `my_lslack ` with something more serious

soccercider committed:
improvement to ecp has been made.

ecp test file added, same as oa but with no 8pp because 8pp is nonconvex and no bounds leads to problems (unless init rNLP is config strategy, then we can use this to get a lower bound).

some termination conditions are different for OA and ECP

some more additions:
-Change names for my_lslack to lower_slack
-Once all tests run, copy them into the test_mindtpy.py (all OA) but with ECP (add the new tests skip 8pp and failing tests).
-Add stalling limit to algorithm_should_terminate terminate with feasible
-Termination condition including feasible (check how to check for two different outputs in assert).
-If we get an error, in the exception catch print name of constraint in logger.

soccercider committed:
-Add breaking conditional if nonlinear constraint has both upper and lower bound. Add logger warning pointing to constraint.

ZedongPeng commented on pyomo/contrib/mindtpy/cut_generation.py:
```diff
@@ -110,6 +110,13 @@ def add_ecp_cuts(target_model, solve_data, config,
+                'and upper bound.'
+                '\n'.format(
+                    constr))
+            break
```
Should we terminate when there is an equality constraint? Or skip and go to the next constraint?

soccercider committed:
removed obj bound from MINLPsimple 1,2,3

soccercider committed:
removed obj bound from MINLPsimple 1,2,3 and add slack

soccercider committed:
modified 8pp to be both nonconvex and convex based on user

soccercider committed:
MindtPy.py: added stalling_limit option config
test_mindtpy_ECP.py: removed int cuts, add bound_tolerances to some

soccercider committed:
cut_generation.py: changed break to continue, info to warning
iterate.py: change feasible to nosolution if stalling
test_mindtpy_ECP.py: added 8pp tests, with obj_bound = 5500

soccercider committed:
min_flag to iterate, comments in cut_generation

added min_flag to see if max or min problem.
added comments to add_ecp-cuts

soccercider committed:
Merge branch 'mindtpy-master' into mindtpy_ecp

soccercider committed:
cut_generation.py: use config.linearize_inactive
iterate.py: remove min_flag since we have objective_sense
test_mindtpy_ECP.py: added maximization test

soccercider committed:
update default for MindtPy.py

ZedongPeng commented on pyomo/contrib/mindtpy/tests/eight_process_problem.py:
```diff
@@ -32,7 +32,7 @@
     """Flowsheet for the 8 process problem."""
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self, convex=True, *args, **kwargs):
```
I don't know if this function works. But I would suggest putting the default arguments behind the non-default arguments.
FYI, https://stackoverflow.com/questions/16932825/why-cant-non-default-arguments-follow-default-arguments

soccercider committed:
8pp.py: putting the default arguments behind the non-default arguments

soccercider committed:
8pp.py: putting the default arguments behind the non-default arguments, REVERTED THIS
